File: conformance/kotlin/IntegrationMutation.template.kt
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject

private val json = Json { ignoreUnknownKeys = true; isLenient = false }

fun main() {
    var ok = true
    val item = json.decodeFromString<OrderItem>("""{"id":"a","quantity":2}""")

    // Mutate quantity → 0 (violates minimum: 1). --mutable makes `element` a
    // var; mutation never re-validates until validate() is called explicitly.
    item.element = JsonObject(item.element.jsonObject + ("quantity" to JsonPrimitive(0)))
    val threw = try {
        item.validate()
        false
    } catch (e: Exception) {
        true
    }
    if (!threw) {
        ok = false
        println("MISMATCH: mutated-invalid instance passed validate()")
    }

    // Mutate back to a valid quantity — validate() passes again.
    item.element = JsonObject(item.element.jsonObject + ("quantity" to JsonPrimitive(2)))
    try {
        item.validate()
    } catch (e: Exception) {
        ok = false
        println("MISMATCH: mutated-valid instance failed validate(): " + e)
    }

    println(if (ok) "PASS" else "FAIL")
}

File: conformance/swift/integration-mutation.template.swift
import Foundation

var ok = true
let data = #"{"id":"a","quantity":2}"#.data(using: .utf8)!
guard var item = try? JSONDecoder().decode(OrderItem.self, from: data) else {
    print("MISMATCH: valid payload failed to decode")
    print("FAIL")
    exit(1)
}
if !item.isValid {
    ok = false
    print("MISMATCH: freshly decoded item should be valid")
}

// Mutate into an invalid state: validation APIs must notice.
item.quantity = 0
if item.isValid {
    ok = false
    print("MISMATCH: quantity=0 should be invalid")
}
if (try? item.validatedJSONData()) != nil {
    ok = false
    print("MISMATCH: validatedJSONData should throw for invalid item")
}

// Mutate back to a valid state: validate() must pass again.
item.quantity = 3
do {
    try item.validate()
} catch {
    ok = false
    print("MISMATCH: validate() after fixing quantity threw \(error)")
}

print(ok ? "PASS" : "FAIL")
